### PR TITLE
Add submit stop

### LINKED ISSUE #893: [Bug]: Run abortion makes an instance unusable
### dstack version

master

### Python version

3.8

### Host OS

macOS

### Host Arch

ARM

### What happened?

1. Run something with `dstack run . -b gcp`
2. Wait till `running` status
3. Press Ctrl+C twice, to abort the run
4. Run something again, to use the same instance
5. Your run is stuck at provisioning

### CLI logs

```shell
...
 Refreshed run shaggy-rattlesnake-1: JobStatus.PROVISIONING
...
```

### Server logs

```shell
DEBUG 2024-02-13T12:08:29.840 dstack._internal.server.services.runner.ssh Cannot connect to 34.172.186.208's API: 409 Client Error: Conflict for url: http://localhost:51158/api/submit
```

### Runner logs

_No response_

### Additional Information

Further investigation showed, that the shim still runs the previous container. Stopping the container (`docker stop`) solved the issue.

Old `abort` relied on instance termination. With pools we should rely on terminating the container from the shim.

Termination without `abort` works fine (but the instance becomes `ready` before the shim deletes the container and is actually ready)

### PR DESCRIPTION
**Pull request by TheBits**
Fix #893

### CODE REVIEW

File: runner/internal/shim/api/server.go [hunk 1]
```diff
@@ -12,6 +12,7 @@ import (
 type TaskRunner interface {
 	Run(context.Context, shim.DockerImageConfig) error
 	GetState() shim.RunnerStatus
+	SubmitStop(bool)
```
Comment (Egor-S) [inline runner/internal/shim/api/server.go]: Why is it named `SubmitStop` and not simply `Stop`?

File: src/dstack/_internal/server/background/tasks/process_finished_jobs.py [hunk 2]
```diff
@@ -90,7 +100,28 @@ async def _process_job(job_id):
...
+            job_model,
+        )
+
         job_model.removed = True
         job_model.last_processed_at = get_current_datetime()
         await session.commit()
         logger.info(*job_log("marked as removed", job_model))
+
+
+@runner_ssh_tunnel(ports=[client.REMOTE_SHIM_PORT, client.REMOTE_RUNNER_PORT])
```
Comment (Egor-S) [inline src/dstack/_internal/server/background/tasks/process_finished_jobs.py]: Do we need REMOTE_RUNNER_PORT there?

File: runner/internal/shim/docker.go [hunk 3]
```diff
@@ -82,10 +85,43 @@ func (d *DockerRunner) Run(ctx context.Context, cfg DockerImageConfig) error {
...
+		timeout := int(0)
+		stopOptions.Timeout = &timeout
+	}
+
+	err := d.client.ContainerStop(ctx, d.currentContainer, stopOptions)
+	if err != nil {
+		log.Printf("Failed to stop container: %s", err)
+	}
+
+	if !d.dockerParams.DockerKeepContainer() {
```
Comment (Egor-S) [inline runner/internal/shim/docker.go]: Isn't it a duplication of `defer` inside of the `Run` method?

File: runner/internal/shim/api/server.go [hunk 4]
```diff
@@ -38,5 +39,6 @@ func NewShimServer(address string, runner TaskRunner, version string) *ShimServe
 	mux.HandleFunc("/api/submit", api.JSONResponseHandler("POST", s.SubmitPostHandler))
 	mux.HandleFunc("/api/healthcheck", api.JSONResponseHandler("GET", s.HealthcheckGetHandler))
 	mux.HandleFunc("/api/pull", api.JSONResponseHandler("GET", s.PullGetHandler))
+	mux.HandleFunc("/api/submit_stop", api.JSONResponseHandler("POST", s.StopPostHandler))
```
Comment (Egor-S) [inline runner/internal/shim/api/server.go]: submit there still seems excessive

File: runner/internal/shim/api/http.go [hunk 5]
```diff
@@ -51,3 +51,20 @@ func (s *ShimServer) PullGetHandler(w http.ResponseWriter, r *http.Request) (int
 		State: string(s.runner.GetState()),
 	}, nil
 }
+
+func (s *ShimServer) StopPostHandler(w http.ResponseWriter, r *http.Request) (interface{}, error) {
+	s.mu.RLock()
+	defer s.mu.RUnlock()
+
```
Comment (Egor-S) [inline runner/internal/shim/api/http.go:58]: Why don't we check the state?

File: runner/internal/shim/api/http.go [hunk 6]
```diff
@@ -51,3 +51,20 @@ func (s *ShimServer) PullGetHandler(w http.ResponseWriter, r *http.Request) (int
...
+	var body SubmitStopBody
+	if err := api.DecodeJSONBody(w, r, &body, true); err != nil {
+		log.Println("Failed to decode submit stop body", "err", err)
+		return nil, err
+	}
+
+	s.runner.Stop(body.Force)
```
Comment (Egor-S) [inline runner/internal/shim/api/http.go:71]: Why don't we show errors to the user, if any?

File: runner/internal/shim/api/schemas.go [hunk 7]
```diff
@@ -8,6 +8,10 @@ type DockerTaskBody struct {
 	ImageName string `json:"image_name"`
 }
 
+type SubmitStopBody struct {
```
Comment (Egor-S) [inline runner/internal/shim/api/schemas.go]: There are still some `SubmitStop` in the code. But I guess it's fine

**Review by Egor-S — APPROVED** [review]

File: runner/internal/shim/api/schemas.go (hunk 7 above)
Comment (TheBits) [reply runner/internal/shim/api/schemas.go]: Fixed.

File: runner/internal/shim/api/http.go (hunk 6 above)
Comment (TheBits) [reply runner/internal/shim/api/http.go:71]: I'll do it in next iteration.

File: runner/internal/shim/api/server.go (hunk 4 above)
Comment (TheBits) [reply runner/internal/shim/api/server.go]: Fixed

File: runner/internal/shim/api/http.go (hunk 5 above)
Comment (TheBits) [reply runner/internal/shim/api/http.go:58]: The state checks in the `DockerRunner.Stop`. I've added the check. But we need to rework this code.

File: runner/internal/shim/docker.go (hunk 3 above)
Comment (TheBits) [reply runner/internal/shim/docker.go]: Yes, it is! Thanks! Fixed.

File: src/dstack/_internal/server/background/tasks/process_finished_jobs.py (hunk 2 above)
Comment (TheBits) [reply src/dstack/_internal/server/background/tasks/process_finished_jobs.py]: Removed

File: runner/internal/shim/api/server.go (hunk 1 above)
Comment (TheBits) [reply runner/internal/shim/api/server.go]: Fixed.

**Pull request merged**